Dev/statement performance analyzer

JesperWisborgKrogh opened the pull request:
Create a report of the statements running on the server.
The views are calculated based on the overall and/or delta activity.

MarkLeith commented on procedures/statement_performance_analyzer.sql:
```diff
@@ -0,0 +1,792 @@
+
+    -- Set configuration options
+    IF (@sys.statement_analyzer_limit IS NULL) THEN
+        SET @sys.statement_analyzer_limit = sys.sys_get_config('statement_analyzer_limit', '100');
```
This should be consistent like the diagnostics routine variables - i.e it should prefix with the routine name like:

statement_performance_analyzer.limit

MarkLeith commented on procedures/statement_performance_analyzer.sql:
```diff
@@ -0,0 +1,792 @@
+            SELECT CONCAT('Top ', @sys.statement_analyzer_limit, ' Queries Using Custom View') AS 'Next Output';
+
+            IF (@sys.statement_analyzer_view IS NULL) THEN
+                SET @sys.statement_analyzer_view = sys.sys_get_config('sys.statement_analyzer_view', NULL);
```
Same here with consistency in naming. And lack of sys_config row for both of the variables too.

MarkLeith commented on README.md:
```diff
@@ -4399,6 +4399,207 @@ mysql> CALL sys.ps_truncate_all_tables(false);
+-- Use a custom view showing the top 10 query sorted by total execution time refreshing the view every minute using
+-- the watch command in Linux.
+
+mysql> mysql> CREATE OR REPLACE VIEW mydb.my_statements AS
```
Odd couple of prompts here

MarkLeith commented on procedures/statement_performance_analyzer.sql:
```diff
@@ -0,0 +1,792 @@
+        END IF;
+        PREPARE stmt_delete FROM @sys.tmp.SQL;
+        EXECUTE stmt_delete;
+        DEALLOCATE PREPARE stmt_delete;
```
This pattern is repeated over and over, it seems like this is one of those cases where we could extract this to something like an "execute_prepared_statement" procedure, and just pass in the SQL string to prepare. I would do something like:

CALL execute_prepared_stmt(CONCAT('DELETE FROM ', v_quoted_table));

And have the routine itself set the input string in to a user variable, check debug, prepare/execute/deallocate.

MarkLeith commented on procedures/statement_performance_analyzer.sql:
```diff
@@ -0,0 +1,792 @@
+            SET v_digests_table = 'tmp_digests_delta';
+            DROP TEMPORARY TABLE IF EXISTS tmp_digests_delta;
+            CREATE TEMPORARY TABLE tmp_digests_delta LIKE tmp_digests;
+            SET @sys.tmp.SQL = CONCAT('INSERT INTO tmp_digests_delta
```
For larger statements like this, the execute_prepared_stmt procedure could still take this user variable (just for readability really), and still just allocate that to another internal user variable.

It also shows that the size of the query could be pretty large too, so make sure that the execute_prepared_stmt procedure *also* nulls out the temp user variable after it's been used.

MarkLeith commented on procedures/statement_performance_analyzer.sql:
```diff
@@ -0,0 +1,792 @@
+                    SET v_custom_name = SUBSTRING(@sys.statement_analyzer_view, CHAR_LENGTH(v_custom_db)+2);
+                END IF;
+
+                CALL table_exists(v_custom_db, v_custom_name, v_custom_view_exists);
```
Prefix this with "sys.table_exists", these can be called from outside of the sys schema.

MarkLeith commented on procedures/statement_performance_analyzer.sql:
```diff
@@ -0,0 +1,792 @@
+
+
+    -- Validate settings
+    CALL table_exists(DATABASE(), 'tmp_digests', v_tmp_digests_table_exists);
```
Prefix this with "sys.table_exists", these can be called from outside of the sys schema.

MarkLeith commented on procedures/statement_performance_analyzer.sql:
```diff
@@ -0,0 +1,792 @@
+               SET MESSAGE_TEXT = v_error_msg;
+        END IF;
+
+        CALL table_exists(v_table_db, v_table_name, v_table_exists);
```
Prefix this with "sys.table_exists", these can be called from outside of the sys schema.

JesperWisborgKrogh commented:
Have made the following changes as requested:

* Renamed @sys.statement_analyzer_limit to @sys.statement_performance_analyzer.limit
* Renamed @sys.statement_analyzer_view to @sys.statement_performance_analyzer.view
* Updated README.md
* Added execute_prepared_stmt() to replace the repeated manual steps
* Prefixed table_exists() with "sys."
